docs/design/cloud-init.rst

```diff
@@ -0,0 +1,353 @@
+should be enough to set the hostname. Alternatively, we could make use of the
+``cc_update_hostname`` module which supports the following keys:
+
+.. code-block:: yaml
+
+  preserve_hostname: <true/false>
+  fqdn: <fqdn>
+  hostname: <fqdn/hostname>
+
+We could ignore the fqdn key and use the other two.
+
+**50ChangeMachineId**: We should probably leave this intact. Newer cloud-init
+version will automatically change the machine ID to a random value as this task
+does, but allowing this task to run to make sure the machine ID is always
+altered even on images that user older versions of cloud-init won't harm.
+
+**50ChangePassword**: This task will change the password and inject SSH
+authorized keys to a list of users defined in the ``USERS`` image property. For
+changing the password of users, we cloud make use of the ``cc_set_password``
+module:
+
+.. code-block:: yaml
+
+    ssh_pwauth: <yes/no/unchanged>
+
+    password: password1
+    chpasswd:
+        expire: <true/false>
+
+    chpasswd:
+        list: |
+            user1:password1
+            user2:RANDOM
+            user3:password3
+            user4:R
+
+The ``password`` key only works for the default user and is not present in
+older versions of cloud-init, which leaves us with the ``chpasswd``. Using this
+key we can define the list of user-password tuples.
+
+Injecting SSH authorized keys to a list of users is not that easy. We can make
+the keys available to cloud-init by either setting the ``public-keys`` metadata
+key or using the ``ssh_authorized_keys`` config key. The ``cc_ssh`` module will
+inject the keys found there to root as well as the default user if defined.
+
+keys In order to make cloud-init work with the
+``USERS`` list that we provide, chpasswd should be used. This task also puts
+ssh authorization keys to the users' home directories. To do this we need to
+set the ``public-keys`` metadata key or use the ``ssh_authorized_keys`` config
+key of ``cc_ssh`` module. The problem is that this will put the keys only to
+the default user and root. You may put ssh keys to other users only when
+defining them using the ``users`` list of the ``cc_users_group`` module:
+
+.. code-block:: yaml
+
+  users:
+    - default
+    - name: <username>
+      expiredate: <date>
+      gecos: <comment>
+      groups: <additional groups>
+      homedir: <home directory>
+      inactive: <true/false>
+      lock_passwd: <true/false>
+      no-create-home: <true/false>
+      no-log-init: <true/false>
+      no-user-group: <true/false>
+      passwd: <password>
+      primary-group: <primary group>
+      selinux-user: <selinux username>
+      shell: <shell path>
+      snapuser: <email>
+      ssh-authorized-keys:
+          - <key>
+          - <key>
+      ssh-import-id: <id>
+      sudo: <sudo config>
+      system: <true/false>
+      uid: <user id>
+
+but we don't want snf-image to overwrite the already defined users or
+read/update existing configuration files.
```

Comment:
Actually something went wrong and this needs rephrasing. The verdict is that in order to change the password to a list of users, we'll need to use `chpasswd` and in order to inject keys we should use the metadata `public-keys` key. The public keys will only be injected to root and the default user if defined. For the vast majority of the cases this is OK.

In the limitations sector we expose the problem that the list of users we should need to feed chpasswd with may change and suggest the introcuction of the users OS parameter.